sklearn/cluster/k_medoids_.py

```diff
@@ -0,0 +1,284 @@
+    array([[1, 2],
+           [4, 2]])
+    >>> kmedoids.inertia_
+    8.0
+
+    References
+    ----------
+    Kaufman, L. and Rousseeuw, P.J., Statistical Data Analysis Based on
+    the L1–Norm and Related Methods, edited by Y. Dodge, North-Holland,
+    405–416. 1987
+    """
+
+    def __init__(self, n_clusters=8, distance_metric='euclidean',
+                 init='heuristic', max_iter=300, random_state=None):
+        self.n_clusters = n_clusters
+        self.distance_metric = distance_metric
+        self.init = init
+        self.max_iter = max_iter
+        self.random_state = random_state
+
+    def _check_init_args(self):
+        """Validates the input arguments. """
+
+        # Check n_clusters
+        if (self.n_clusters is None or self.n_clusters <= 0 or
+                not isinstance(self.n_clusters, (int, np.integer))):
+            raise ValueError("n_clusters should be a nonnegative integer. "
+                             "%s was given" % self.n_clusters)
+
+        # Check init
+        init_methods = ['random', 'heuristic']
+        if self.init not in init_methods:
+            raise ValueError("init needs to be one of " +
+                             "the following: " +
+                             "%s" % init_methods)
+
+        # Check random state
+        self.random_state_ = check_random_state(self.random_state)
+
+    def fit(self, X, y=None):
+        """Fit K-Medoids to the provided data.
+
+        Parameters
+        ----------
+        X : {array-like, sparse matrix}, shape = (n_samples, n_features).
+            Dataset to cluster.
+
+        Returns
+        -------
+        self
+        """
+
+        self._check_init_args()
+        X = check_array(X, accept_sparse=['csr', 'csc'])
+        if self.n_clusters > X.shape[0]:
+            raise ValueError("The number of medoids %d must be less "
+                             "than the number of samples %d."
+                             % (self.n_clusters, X.shape[0]))
+
+        distances = pairwise_distances(X, metric=self.distance_metric)
+
+        medoid_idxs = self._get_initial_medoid_indices(distances,
+                                                       self.n_clusters)
+        labels = np.argmin(distances[medoid_idxs, :], axis=0)
+
+        # Old medoids will be stored here for reference
+        old_medoid_idxs = np.zeros((self.n_clusters,))
+
+        # Continue the algorithm as long as
+        # the medoids keep changing and the maximum number
+        # of iterations is not exceeded
+        self.n_iter_ = 0
+        while (not np.all(old_medoid_idxs == medoid_idxs) and
+                self.n_iter_ < self.max_iter):
+            self.n_iter_ += 1
+            old_medoid_idxs = np.copy(medoid_idxs)
+            labels = np.argmin(distances[medoid_idxs, :], axis=0)
+
+            # Update medoids with the new cluster indices
+            self._update_medoid_idxs_in_place(distances, labels, medoid_idxs)
+
```

Comment:
break here upon convergence. also add an `else` to the for loop for the non-convergence case to raise a ConvergenceWarning

Reply:
done with warnings.warn